neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -246,17 +257,93 @@ func QmpQueryMemoryDevices(ip string, port int32) ([]QmpMemoryDevice, error) {
 	return result.Return, nil
 }
 
+func QmpMatchName(reg *regexp.Regexp, name string) (int, bool) {
+	// extracts the id from the name
+	matches := reg.FindStringSubmatch(name)
+	if len(matches) < 2 {
+		return 0, false
+	}
+	id, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return 0, false
+	}
+	return id, true
+}
+
+var memBackendRegexp = regexp.MustCompile(`memslot(\d+)`)
+
+func QmpMatchMemBackend(name string) (int, error) {
+	// "/objects/memslot3" -> 3
+	name = strings.ReplaceAll(name, "/objects/", "")
+	idx, ok := QmpMatchName(memBackendRegexp, name)
+	if !ok {
+		return 0, fmt.Errorf("failed to parse memory device id: %s", name)
+	}
+	return idx, nil
+}
+
+func QmpQueryMemoryBackendIds(mon *qmp.SocketMonitor) (map[int]struct{}, error) {
+	cmd := []byte(`{"execute": "qom-list", "arguments": {"path": "/objects"}}`)
+	raw, err := mon.Run(cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	var result QmpObjects
+	json.Unmarshal(raw, &result)
+	backends := map[int]struct{}{}
+	for _, o := range result.Return {
+		idx, err := QmpMatchMemBackend(o.Name)
+		if err == nil {
+			backends[idx] = struct{}{}
+		}
+	}
+	return backends, nil
+}
+
+func QmpAddMemoryBackend(mon *qmp.SocketMonitor, idx int, size int64) error {
```

Comment:
Needs documentation:

- What does this function do?
- what's the units of `size`?
- what's the relationship between this function and `QmpAddMemoryDevice`? (which one goes first?)

I happen to know the answers here. I ask because the comments should give a reasonable overview for someone who's not intimately familiar with QEMU — otherwise we risk limiting the people who can work on neonvm-controller to those familiar with both QEMU and Kubernetes.

I would imagine something like:

> QmpAddMemoryBackend adds a single memory slot to the VM with the given size.
>
> The memory slot does nothing until a corresponding "device" is added to the VM for the same memory slot. See QmpAddMemoryDevice for more.
> When unplugging, QmpDelMemoryDevice must be called before QmpDelMemoryBackend.
>
> The size is in units of bytes.

The other functions below (`Qmp{Add,Del}Memory{Backend,Device}`) should have similar comments

Reply:
I suggest going with the common practice in Go. For exported types, functions, etc, you should have a comment providing necessary the documentation (details [here](https://tip.golang.org/doc/comment)). The comment should start with the name of the exported entity.

Reply:
We can add docs here.